trie/stacktrie.go

```diff
@@ -0,0 +1,263 @@
+// used to calculate the hash of a trie. The core idea is that at
+// any time, only one branch is expanded and the rest is hashed as
+// soon as it is determined it is no longer needed.
+type StackTrie struct {
+	stack  []StackTrieItem
+	top    int
+	hasher *hasher
+}
+
+// NewStackTrie builds a new stack trie. The whole stack space is
+// pre-allocated so as to save reallocations down the road.
+func NewStackTrie() *StackTrie {
+	return &StackTrie{
+		top:    -1,
+		stack:  make([]StackTrieItem, 65),
+		hasher: newHasher(false),
+	}
+}
+
+func (st *StackTrie) TryUpdate(key, value []byte) error {
+	k := keybytesToHex(key)
+	if len(value) == 0 {
+		panic("deletion not supported")
+	}
+	st.insert(&st.stack[0].ext, nil, k, valueNode(value))
+	//fmt.Println("trie=", &st.stack[0].ext)
+	return nil
+}
+
+// alloc prepares the next stage in the stack for reuse.
+func (st *StackTrie) alloc() {
+	for i := 0; i < 16; i++ {
+		st.stack[st.top+1].branch.Children[i] = nil
+	}
+
+	st.top++
+}
+
+func (st *StackTrie) insert(n node, prefix, key []byte, value node) node {
+	// Special case: the trie is empty
+	if st.top == -1 {
+		st.top = 0
+		st.stack[st.top].depth = 0
+		st.stack[st.top].ext.Key = key
+		st.stack[st.top].ext.Val, _ = st.hasher.hash(value, false)
+		st.stack[st.top].keyUntilHere = []byte("")
+
+		return &st.stack[st.top].ext
+	}
+
+	// Use the prefix key to find the stack level in which the code needs to
+	// be inserted.
+	level := -1
+	for index := st.top; index >= 0; index-- {
+		level = index
+		if bytes.Equal(st.stack[level].keyUntilHere, key[:len(st.stack[level].keyUntilHere)]) {
+			// Found the common denominator, stop the search
+			break
+		}
+	}
+
+	// Already hash the value, which it will be anyway
+	hv, _ := st.hasher.hash(value, false)
+
+	// The difference happens at this level, find out where
+	// exactly. The extension part of the fullnode part?
+	extStart := len(st.stack[level].keyUntilHere)
+	extEnd := extStart + len(st.stack[level].ext.Key)
+	if bytes.Equal(st.stack[level].ext.Key, key[extStart:extEnd]) {
+		// The extension and the key are identical on the length of
+		// the extension, so st.stack[level].ext.Val should point to
+		// st.stack[level].branch, and the difference should be foud
+		// there.
+		var fn *fullNode
+		fn = &st.stack[level].branch
+
+		// The correct entry is the only one that isn't nil
+		for i := 15; i >= 0; i-- {
+			if fn.Children[i] != nil {
+				switch fn.Children[i].(type) {
+				// Only hash entries that are not already hashed
+				case *fullNode, *shortNode:
+					fn.Children[i], _ = st.hasher.hash(fn.Children[i], false)
+					st.top = level
+				default:
+				}
+				break
+			}
+		}
+
+		// That fullNode should have at most one non-hashNode child,
+		// hash it because no more nodes will be inserted in it.
+		st.alloc()
+		keyUntilHere := len(st.stack[level].keyUntilHere) + len(st.stack[level].ext.Key) + 1
+		st.stack[level].branch.Children[key[keyUntilHere-1]] = &st.stack[st.top].ext
```

Comment:
please hide the stack internals, and use ops like `stack.push`/`pop()`/`peek()`

Reply:
This is going to be removed soon